### DISCUSSION TITLE
General Confusion About Keys

### DISCUSSION THREAD

**Discussion by alatark** [rustdesk/rustdesk#7650] [General]
Sorry for such a dumb question, but I'm finding all the references to "Key" confusing.

I added the app to two test machines here and got them working using the public server.

I then installed a server on a third machine in the local subnet. The installation did, as promised, make private and public keys in a pair of id_ed25519 files.

I changed the ID and Relay servers in both machines to point to local server, and restarted the service, but my attempts to connect between the two clients all yield "Connection Error: Key Mismatch".

I can understand this, as I never filled in a key, but what do I fill in? The public key of my local server? The private key? An invented key that only needs to match at the two ends? Some other key I'm not aware of?

It says "Key Mismatch", but is this a mismatch between the keys in the two clients? Or between one client and the server? Or even between the public key and the private key?

Neither signalserver.log nor relayserver.log get any lines added when this failed connection is attempted, so I'm not even sure either client is actually connecting to the server at all.

Thanks in advance for all help,
Paul

-----------------------------------------

PS: my original executables on each client were downloaded from the cloud, but the server setup did successfully publish install scripts for linux and windows on port 8000. I tried reinstalling both clients using the powershell script from the installed server, but there was no change in behaviour. However, I notice the settings were retained across the reinstallation, so perhaps not everything was correctly reset when the app was reinstalled?

**Comment by alatark** [comment]
This entire sorry issue was entirely my own fault, and is now resolved. Somehow I had not noticed that there was a second section (2.3) in the instructions discussing client configuration (in addition to section 1).

I must have wound up trying a dozen different combinations before it started working, and I believe I even tested the right combination several times during my flailing about, albeit without success. I suspect this was because I failed to do sufficient service and/or machine restarts at the time. In the end I simply powered all three machines off and back on; then it worked.

What worked for me (testing within a single subnet, with two Windows clients and a Ubuntu server):
- in both the local and remote clients, use the local IP address of the server in **ID server** and leave **Relay server** blank;
- in the **key** field, use the public key generated by the server install (id_ed25519.pub), truncated after the first "=";
- stop and restart the "Rustdesk service" on both clients (or just restart them);
- restart the hbbs and hbbr services on the server (in my case called rustdesksignal.service and rustdeskrelay.service), or simply reboot the server.

Then it worked! I'm sure there are other, probably simpler ways, but this way at least works for me.

**Comment by xlionjuan** [comment]
Related search keyword: Asymmetric cryptography

You need `id_ed25519.pub` (public key) to put on your client, and please don't modify key files manually, no matter it is public key or private key 

Sometime, the server may generate invalid key, please delete them and let it regenerate 


And make sure both your `hbbr` and `hbbs`'s execute parm has `-k _` at the end.

**Reply by KevinNeale** [reply]
Can you please expand on "And make sure both your hbbr and hbbs's execute parm has -k _ at the end."? Where do we put this -k?

**Reply by xlionjuan** [reply]
`-k` parameter isn't needed anymore.

**Comment by daparix** [comment]
In linux, the key file is in /var/lib/rustdesk-server, it is the one with the *.pub extension.
